FullStorySegmentMiddleware/Classes/FullStoryMiddleware.m

```diff
@@ -125,102 +127,148 @@ - (SEGPayload *) getNewPayloadWithFSURL:(SEGContext * _Nonnull)context {
+- (NSDictionary *) getSuffixedProps: (NSDictionary *)properties {
     //TODO: Segment will crash and not allow props to have curcular dependency, but we should handle it here anyways
     
     NSMutableDictionary *props = [[NSMutableDictionary alloc] initWithCapacity:[properties count]];
-    for(id key in properties){
+    for(NSString *key in properties){
         NSString *suffix = @"";
 
-        // properties should always be a NSDictionary, but still check to make sure the value is an object
+        // properties should always be a NSDictionary, but still check to make sure the value we get is an object
         // Check Type Encoding: https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/ObjCRuntimeGuide/Articles/ocrtTypeEncodings.html
-        const char* typeCode = @encode(typeof([props valueForKey:key]));
+        const char* typeCode = @encode(typeof(props[key]));
 
-        if([@"@" isEqualToString:@(typeCode)]){
-            NSObject *obj = [properties valueForKey:key];
+        if ([@"@" isEqualToString:@(typeCode)]) {
+            NSObject *obj = properties[key];
             
-            // if it's a NSDictionary: recurrsively get props, and we can not take underscore in key for nested objects, so we just remove them
+            // if it's a NSDictionary: recurrsively get props, and we can not take underscore in key for nested objects
             // more info: https://help.fullstory.com/hc/en-us/articles/360020623234-FS-Recording-Client-API-Requirements
-            if([obj isKindOfClass:[NSDictionary class]]){
-                [props setValue:[self getSuffixedProps:[properties valueForKey:key]]
+            if ([obj isKindOfClass:[NSDictionary class]]) {
+                [props setValue:[self getSuffixedProps:properties[key]]
                          forKey:[key stringByReplacingOccurrencesOfString:@"_" withString:@""]];
-            }else if ([obj isKindOfClass:[NSArray class]]){
+            } else if ([obj isKindOfClass:[NSArray class]]) {
                 // array of dicts? array of strings/ints/numbers? yuck!
-                NSMutableArray *arr = (NSMutableArray *) obj;
-                NSDictionary *dict = [self getObjectFromArrayObject:arr withKey:key];
+                NSArray *arr = (NSArray *) obj;
+                NSDictionary *dict = [self getDictionaryFromArrayObject:arr withKey:key];
                 [props addEntriesFromDictionary:dict];
-            }else{
+            } else {
+                // if this is not a dictionary or array, then treat it like simple values, if data falls outside of these, we don't try to infer anything and just send as is to the server
                 suffix = [self getSuffixStringFromSimpleObject:obj];
-                [props setValue:[properties valueForKey:key] forKey:[key stringByAppendingString:suffix]];
+                [self appendToDictionary:props withKey:[key stringByAppendingString:suffix] andValue:properties[key]];
```

Comment:
`properties[key]` has already been stored in `obj` so you can use `andValue:obj`